### PR TITLE
Improve `AdtDef` interning.

### PR DESCRIPTION
**Pull request by nnethercote**
This commit makes `AdtDef` use `Interned`. Much of the commit is tedious
changes to introduce getter functions. The interesting changes are in
`compiler/rustc_middle/src/ty/adt.rs`.

r? @fee1-dead

### CODE REVIEW

**Comment by fee1-dead** [comment]
Since `AdtDef` already uses its `DefId` as a hash I don't think there is going to be any performance impact here, but it is nice to be safe.

@bors try @rust-timer queue

**Comment by camsteffen** [comment]
These interning changes are neat.

File: compiler/rustc_middle/src/ty/adt.rs [hunk 1]
```diff
@@ -89,62 +90,62 @@ bitflags! {
 /// where `x` here represents the `DefId` of `S.x`. Then, the `DefId`
 /// can be used with [`TyCtxt::type_of()`] to get the type of the field.
 #[derive(TyEncodable, TyDecodable)]
-pub struct AdtDef {
+pub struct AdtDefS {
```
Comment (petrochenkov) [inline compiler/rustc_middle/src/ty/adt.rs:93]: Can the naming convention be changed to something like `AdtDefData`? (Like e.g. `Span` and `SpanData`.)

I know that the `S` comes from `TyS`, but `TyS` doesn't look like an intentional choice, but rather like some horrible legacy naming that nobody likes but has no time to rename.

Comment (nnethercote) [reply compiler/rustc_middle/src/ty/adt.rs]: I have updated the code to use `AdtData`.

**Comment by nnethercote** [comment]
I have fixed the conflicts. I also introduced `AdtDef::variant()` and rewrote a bunch of `def.variants()[n]` expressions as `def.variant(n)`.

**Review by fee1-dead — APPROVED** [review]
LGTM, Thanks!

**Pull request merged**